core/impl/src/main/java/com/blazebit/persistence/impl/function/window/AbstractWindowFunction.java

```diff
@@ -0,0 +1,557 @@
+            render(context, frame.getStart());
+            context.addChunk("AND ");
+            render(context, frame.getEnd());
+        }
+    }
+
+    protected void render(FunctionRenderContext context, FramePosition position) {
+        switch (position.getType()) {
+            case UNBOUNDED_PRECEDING:
+                context.addChunk("UNBOUNDED PRECEDING");
+                break;
+            case BOUNDED_PRECEDING:
+                context.addChunk(position.getExpression());
+                context.addChunk(" PRECEDING");
+                break;
+            case CURRENT_ROW:
+                context.addChunk("CURRENT ROW");
+                break;
+            case UNBOUNDED_FOLLOWING:
+                context.addChunk("UNBOUNDED FOLLOWING");
+                break;
+            case BOUNDED_FOLLOWING:
+                context.addChunk(position.getExpression());
+                context.addChunk(" FOLLOWING");
+                break;
+            default:
+                throw new IllegalArgumentException("No branch for " + position.getType());
+        }
+        context.addChunk(" ");
+    }
+
+    protected void appendEmulatedOrderByElementWithNulls(FunctionRenderContext context, Order element) {
+        context.addChunk("case when ");
+        context.addChunk(element.getExpression());
+        context.addChunk(" is null then ");
+        context.addChunk(element.isNullsFirst() ? "0" : "1");
+        context.addChunk(" else ");
+        context.addChunk(element.isNullsFirst() ? "1" : "0");
+        context.addChunk(" end, ");
+        context.addChunk(element.getExpression());
+        context.addChunk(element.isAscending() ? " asc" : " desc");
+    }
+
+
+    /**
+     *
+     * @author Jan-Willem Gmelig Meyling
+     * @author Sayra Ranjha
+     * @since 1.4.0
+     */
+    private enum Mode {
+        ARGUMENTS,
+        PARTITION_BY,
+        ORDER_BY,
+        FRAME_CLAUSE
+    }
+    
+    /**
+     *
+     * @author Jan-Willem Gmelig Meyling
+     * @author Sayra Ranjha
+     * @since 1.4.0
+     */
+    protected static final class WindowFunction {
+
+        private String functionName;
+
+        private final List<String> arguments = new ArrayList<>();
+        
+        private final List<String> filterExpressions = new ArrayList<>();
+
+        private final WindowDefinition window = new WindowDefinition();
+
+        public String getFunctionName() {
+            return functionName;
+        }
+
+        public List<String> getArguments() {
+            return arguments;
+        }
+
+        public List<String> getFilterExpressions() {
+            return filterExpressions;
+        }
+
+        public WindowDefinition getWindow() {
+            return window;
+        }
+    }
+
+    /**
+     *
+     * @author Jan-Willem Gmelig Meyling
+     * @author Sayra Ranjha
+     * @since 1.4.0
+     */
+    protected static final class WindowDefinition {
+
+        private Partition partition = new Partition();
```

Comment:
Inline the array list

Reply:
Are you sure? I think for custom extensions the abstraction could really ease translating the structure  to a new SQL dialect and which exceptions to keep in mind. See for example how I broke out the render method in individual protected methods that can be overridden.

Reply:
You can still have separate methods that accept a subset of the window fields to do the rendering. The intention is to avoid unnecessary objects i.e. flatten them away.